processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java

```diff
@@ -408,7 +354,174 @@ void addPreFilter(@Nullable final Filter filter)
+   * B. Unnesting an expression from multiple columns e.g. select * from foo, UNNEST(ARRAY[dim1,dim2]) as u(d12)
+   * In case A, d3 is a direct reference to dim3 so any filter using d3 can be rewritten using dim3 and added to pre filter
+   * while in case B, due to presence of the expression virtual column expressionVirtualColumn("j0.unnest", "array(\"dim1\",\"dim2\")", ColumnType.STRING_ARRAY)
+   * the filters on d12 cannot be pushed to the pre filters
+   *
+   * @param queryFilter            query filter passed to makeCursors
+   * @param inputColumn            input column to unnest if it's a direct access; otherwise null
+   * @param inputColumnCapabilites input column capabilities if known; otherwise null
+   */
+  private List<Filter> recursiveRewriteOnUnnestFilters(
+      BooleanFilter queryFilter,
+      final String inputColumn,
+      final ColumnCapabilities inputColumnCapabilites,
+      final FilterSplitter filterSplitter,
+      final boolean isTopLevelAndFilter
+  )
+  {
+    final List<Filter> preFilterList = new ArrayList<>();
+    for (Filter filter : queryFilter.getFilters()) {
+      if (filter.getRequiredColumns().contains(outputColumnName)) {
+        if (filter instanceof AndFilter) {
+          preFilterList.add(new AndFilter(recursiveRewriteOnUnnestFilters(
+              (BooleanFilter) filter,
+              inputColumn,
+              inputColumnCapabilites,
+              filterSplitter,
+              isTopLevelAndFilter
+          )));
+        } else if (filter instanceof OrFilter) {
+          // in case of Or Fiters, we set isTopLevelAndFilter to false that prevents pushing down any child filters to base
+          List<Filter> orChildFilters = recursiveRewriteOnUnnestFilters(
+              (BooleanFilter) filter,
+              inputColumn,
+              inputColumnCapabilites,
+              filterSplitter,
+              false
+          );
+          preFilterList.add(new OrFilter(orChildFilters));
+        } else {
+          final Filter newFilter = rewriteFilterOnUnnestColumnIfPossible(
+              filter,
+              inputColumn,
+              inputColumnCapabilites
+          );
+          if (newFilter != null) {
+            // this is making sure that we are not pushing the unnest columns filters to base filter without rewriting.
+            preFilterList.add(newFilter);
+            filterSplitter.addToPreFilterCount(1);
+          }
+          /*
+           Push down the filters to base only if top level is And Filter
+           we can not push down if top level filter is OR or unnestColumn is derived expression like arrays
+           */
+          if (isTopLevelAndFilter && getUnnestInputIfDirectAccess(unnestColumn) != null) {
+            filterSplitter.addPreFilter(newFilter != null ? newFilter : filter);
+            filterSplitter.addToPreFilterCount(1);
+          }
+          filterSplitter.addToOriginalFilterCount(1);
```

Comment:
Isn't the filter getting added to the preFilter list twice ? It already got added to preFilterList once in line 503. This would cause the count to go up and when you make the comparison of count on line 330, it would give the wrong count.

Reply:
fixed